src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,189 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
+    public MyLinkedList() {
+        this.first = null;
+        this.last = null;
+        this.size = 0;
+    }
+
+    private void validIndex(int index) {
+        if (index < 0 || index >= this.size) {
+            throw new ArrayIndexOutOfBoundsException("Illegal Index: " + index);
+        }
+    }
+
+    private Node<T> indexedNode(int index) {
+        Node<T> node;
+        if (index < this.size / 2) {
+            node = this.first;
+            for (int i = 0; i < index; i++) {
+                node = node.nextNode;
+            }
+            return node;
+        }
+        node = this.last;
+        for (int i = this.size - 1; i > index; i--) {
+            node = node.prevNode;
+        }
+        return node;
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == 0) {
+            Node<T> node0 = new Node<>(value);
+            this.first = node0;
+            this.last = node0;
+            this.size++;
+            return;
+        }
+        Node<T> node = new Node<>(last, null, value);
+        this.last.nextNode = node;
+        this.last = node;
+        this.size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == 0 && this.size == 0) {
+            this.add(value);
+            return;
+        }
+        if (index == 0) {
+            Node<T> newNode = new Node<>(null, this.first.nextNode, value);
+            this.first.prevNode = newNode;
+            this.first = newNode;
+            this.size++;
+            return;
+        }
+        this.validIndex(index - 1);
+        if (index != this.size) {
+            Node<T> node = this.indexedNode(index);
+            Node<T> newNode = new Node<>(node.prevNode, node.nextNode, value);
+            node.prevNode = newNode;
+            node = newNode.prevNode;
+            node.nextNode = newNode;
+            this.size++;
+            return;
+        }
+        this.add(value);
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (T i : list) {
```

Comment:
`i` isn't very informative name of the variable

Reply:
done